client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx

```diff
@@ -0,0 +1,173 @@
+import { addEmailForward } from 'calypso/state/email-forwarding/actions';
+import getEmailForwardingLimit from 'calypso/state/selectors/get-email-forwarding-limit';
+import getEmailForwardingType from 'calypso/state/selectors/get-email-forwarding-type';
+import { getEmailForwards, isAddingForward } from 'calypso/state/selectors/get-email-forwards';
+import { getSiteSlug } from 'calypso/state/sites/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+class EmailForwardingAddNewCompactList extends Component {
+	static propTypes = {
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		selectedDomainName: PropTypes.string.isRequired,
+	};
+
+	state = {
+		forwards: [ { destination: '', mailbox: '', valid: false } ],
+	};
+
+	hasForwards() {
+		return this.props.emailForwards.length > 0;
+	}
+
+	hasReachedLimit() {
+		return this.props.emailForwards.length >= this.props.emailForwardingLimit;
+	}
+
+	validForwards() {
+		return ! this.state.forwards.some( ( t ) => ! t.valid );
+	}
+
+	addNewEmailForwardWithAnalytics = ( domainName, mailbox, destination, siteSlug ) => {
+		const onSuccessRedirectTarget = () => {
+			page( emailManagement( siteSlug, domainName ) );
+		};
+
+		withAnalytics(
+			composeAnalytics(
+				recordTracksEvent(
+					'calypso_domain_management_email_forwarding_add_new_email_forward_click',
+					{
+						destination,
+						domain_name: domainName,
+						mailbox,
+					}
+				),
+				this.props.addEmailForward( domainName, mailbox, destination, onSuccessRedirectTarget )
+			)
+		);
+	};
+
+	addNewEmailForwardsClick = ( event ) => {
+		const { selectedSiteSlug, addingForward } = this.props;
+
+		event.preventDefault();
+
+		if ( addingForward ) {
+			return;
+		}
+
+		this.state.forwards.map( ( t ) => {
+			const { mailbox, destination } = t;
+
+			this.addNewEmailForwardWithAnalytics(
+				this.props.selectedDomainName,
+				mailbox,
+				destination,
+				selectedSiteSlug
+			);
+		} );
+	};
+
+	onForwardAdd = () => {
+		this.setState( { forwards: [ ...this.state.forwards, { destination: '', mailbox: '' } ] } );
+	};
+
+	renderAddButton( addMoreButton = false ) {
+		const { translate } = this.props;
+		return (
+			<div className="email-forwarding-add-new-compact-list__actions">
+				{ addMoreButton && (
+					<Button
+						className="email-forwarding-add-new-compact-list__add-another-forward-button"
+						onClick={ this.onForwardAdd }
+					>
+						<Gridicon icon="plus" />
+						<span>{ translate( 'Add another forward' ) }</span>
+					</Button>
+				) }
+
+				<Button
+					primary
+					onClick={ this.addNewEmailForwardsClick }
+					disabled={ ! this.validForwards() }
+				>
+					{ translate( 'Add' ) }
+				</Button>
+			</div>
+		);
+	}
+
+	renderFormFooter() {
+		return <div>{ this.renderAddButton() }</div>;
+	}
+
+	removeHandler = ( index ) => {
+		const array = this.state.forwards;
+		array.splice( index, 1 );
+
+		this.setState( { forwards: array } );
```

Comment:
I think destructuring could help here (and in the other handler):

```suggestion
		const [ forwards ] = this.state;
		forwards.splice( index, 1 );

		this.setState( { forwards } );
```